### PR TITLE
[TF-TRT] Add Layout + ConstFold optimization on the TRT engine op

### PR DESCRIPTION
**Pull request by Kh4L**
### Motivation
The introduction of Tensor Cores in recent GPU architectures (and the subsequent changes in cuDNN) made `NHWC` the preferred layout for convolutions in TF.

However, the TensorRT API requires the input for convolution to be in `NCHW`. To satisfy this requirement, TF-TRT was inserting transpose ops around convolutions when necessary.

This PR adds a layout optimization pass with a target layout of `NCHW`, which ensures that most of Conv kernels will be in TRT compatible layout. Additionally, the layout optimization pass can eliminate redundant layout transpose operations. This results in a simpler TRT network with a larger potential for optimization.

### Description
TF-TRT now accomplishes it by running both a grapplers's `GenericLayoutOptimizer` followed by a ConstantFolding optimizers  in the `ConvertGraphDefToEngine`, applying them on a segment level.

The LayoutOptimizer ensures that the number of transposition is optimal. This optimization is described in slides 18-21 of [TensorFlow Graph Optimizations presentation ](https://web.stanford.edu/class/cs245/slides/TFGraphOptimizationsStanford.pdf)

This PR also modifies the `GenericLayoutOptimizer` to enforce a specific layout is set as layout target, through a ctor parameter.

The layout optimizer can be enabled with the environment variable `TF_TRT_ENABLE_LAYOUT_OPTIMIZER` (default=1, set to 0 to disable).

### Dependency

This PR needs the changes in https://github.com/tensorflow/tensorflow/pull/53368

### CODE REVIEW

File: tensorflow/core/grappler/optimizers/generic_layout_optimizer.cc
Comment (bixia1) [inline tensorflow/core/grappler/optimizers/generic_layout_optimizer.cc:116]: Let's move this comment line into the else-branch, and revise it to:

// Otherwise, enforced_layout is NCHW enforced by previous check.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:1292]: Would you please create a github ticket and describe the TODO there, then use the ticket number here as TODO(path to github ticket)?

**Comment by bixia1** [comment]
Looks good to me! Would you please split the grappler change to a different PR?

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6813]: Note that constant folding can affect how we handle quantization ops. See note here https://github.com/tensorflow/tensorflow/blob/afec6f4d62c1ba96b3899d48d35793dfeb85ab6e/tensorflow/compiler/tf2tensorrt/trt_convert_api.cc#L93-L100

You can use the last arg of ConstantFolding to control folding of quantization ops: https://github.com/tensorflow/tensorflow/blob/afec6f4d62c1ba96b3899d48d35793dfeb85ab6e/tensorflow/core/grappler/optimizers/constant_folding.h#L50

**Review by tfeher — CHANGES_REQUESTED** [review]
Thanks Serge for the PR! In general it looks good, but there is a potential interaction with PR #52248 that should be addressed.

File: tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc
Comment (christopherbate) [inline tensorflow/compiler/tf2tensorrt/kernels/trt_engine_op.cc:1019]: Perhaps sanity check the pointers, we have macros for that:
```
TRT_ENSURE(cache_resource);
TRT_ENSURE(ctx);
```

**Comment by christopherbate** [comment]
How can we force disabling of this layout optimizer if needed? Add an option to the TRTEngineOp? Or read environment variable?

File: tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc
Comment (Kh4L) [reply tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:721]: Because `params.cluster` is const while `TRTOptimizationPass::Optimize` takes a non-const pointer.
Making `params.const` non-const would be also an option.

**Comment by Kh4L** [comment]
Following @christopherbate comment above, we had an internal discussion on wether to make the feature optional or not.

As per @tfeher comment
"Since we are modifying the graph after segmentation, we could argue that we are on dangerous territory. That would justify tagging it as experimental.", the best option would be to leave the optimization pass default, but add a env var to disable it.

@bixia1 what are you thoughts on this?

**Comment by bixia1** [comment]
> Following @christopherbate comment above, we had an internal discussion on wether to make the feature optional or not.
> 
> As per @tfeher comment "Since we are modifying the graph after segmentation, we could argue that we are on dangerous territory. That would justify tagging it as experimental.", the best option would be to leave the optimization pass default, but add a env var to disable it.
> 
> @bixia1 what are you thoughts on this?

Yes, agree.

**Comment by Kh4L** [comment]
The grappler changes have been moved to https://github.com/tensorflow/tensorflow/pull/53368 and the optimizer can be now be enabled/disabled with the envvar `TF_TRT_ENABLE_LAYOUT_OPTIMIZER`.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6857]: Note that we have a helper function, which might be more robust: https://github.com/tensorflow/tensorflow/blob/781bac10156001569c5e99057f6581cd38773fc6/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L2349-L2355

**Review by tfeher — COMMENTED** [review]
@Kh4L Thank you for updating the PR! I have one more comment

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6849]: This is not needed.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6852]: Not needed either.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6857]: You can use ReadBoolFromEnvVar like [here](https://github.com/sourcecode369/tensorflow-1/blob/9f446aba8aaeb2b3c4c6e5ba1ab4cf31494b8a64/tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc#L2360).

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6850]: This is the graph we will convert to engine. Let's just call it "graph".

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6862]: Let's put this comment here:
// TensorRT API requires the input for convolution to be in NCHW.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h:162]: We can make cluster=nullptr as default to avoid the change to the test.

Let's add this to the comment block for this routine:
When 'cluster' is not null, it contains the graph to be converted. We may perform additional optimizations to the graph before converting the graph.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6857]: If cluster is nullptr, we don't even need to read the env. Can you fix this?
Please also squash your commits into one. I will approve this after the layout PR is merged.

**Comment by gbaned** [comment]
@Kh4L Can you please check @bixia1's comments and keep us posted ? Thanks!

File: tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc
Comment (Kh4L) [reply tensorflow/compiler/tf2tensorrt/convert/convert_graph.cc:721]: You are right, I refactored it as suggested in the latest commit

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (Kh4L) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Fixed

**Comment by tfeher** [comment]
@bixia1 could you have a look at the PR? I believe all the issues have been addressed by @Kh4L.

**Review by tfeher — APPROVED** [review]
Thanks @Kh4L for addressing the issues. The PR looks good to me.

**Review by bixia1 — APPROVED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6862]: Need to handle the return value from the call.

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (bixia1) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6869]: Siimlar to the above, need to handle the return value from the call.

**Review by bixia1 — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (tfeher) [inline tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc:6701]: Let's change the error checking here (`TF_CHECK_OK` would abort the process in case of error, which is unnecessary).
```suggestion
      TF_RETURN_IF_ERROR(
```

File: tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc
Comment (Kh4L) [reply tensorflow/compiler/tf2tensorrt/convert/convert_nodes.cc]: Done

**Review by bixia1 — APPROVED** [review]

**Review by bixia1 — APPROVED** [review]

**Pull request merged**
